crates/oxc_linter_plugin/src/plugin.rs

```diff
@@ -0,0 +1,615 @@
+    }
+
+    let program = allocator.alloc(ret.program);
+    let SemanticBuilderReturn { semantic, errors } =
+        SemanticBuilder::new(source_text, source_type).with_trivias(ret.trivias).build(program);
+
+    if !errors.is_empty() {
+        return Err(errors);
+    }
+
+    let semantic = Rc::new(semantic);
+
+    let mut lint_ctx = LintContext::new(&Rc::clone(&semantic));
+
+    let result = plugin.run_tests(
+        &mut lint_ctx,
+        test.relative_path.iter().map(|el| Some(el.clone())).collect::<Vec<_>>(),
+        RulesToRun::Only(rule_name.to_string()),
+        mini_v8,
+    );
+
+    if let Some(err) = result.err() {
+        return Err(vec![err]);
+    }
+
+    Ok(lint_ctx.into_message().into_iter().map(|m| m.error).collect::<Vec<_>>())
+}
+
+#[derive(Debug, Error, Diagnostic)]
+#[error("Test expected to pass, but failed.")]
+struct ExpectedTestToPassButFailed {
+    #[source_code]
+    query: NamedSource,
+    #[label = "This test failed."]
+    err_span: SourceSpan,
+    #[related]
+    errors: Vec<Report>,
+}
+
+#[derive(Debug, Error, Diagnostic)]
+#[error("Test expected to fail, but passed.")]
+struct ExpectedTestToFailButPassed {
+    #[source_code]
+    query: NamedSource,
+    #[label = "This test should have failed but it passed."]
+    err_span: SourceSpan,
+}
+
+#[derive(Debug, Error, Diagnostic)]
+#[error("Unexpected errors in fail test.")]
+struct UnexpectedErrorsInFailTest {
+    #[related]
+    errors: Vec<Report>,
+    #[source_code]
+    query: NamedSource,
+    #[label = "This test should have failed but it passed."]
+    err_span: SourceSpan,
+}
+
+enum PassOrFail {
+    Pass,
+    Fail,
+}
+
+fn span_of_test_n(
+    yaml_text: &str,
+    test_ix: usize,
+    test_code: &str,
+    pass_or_fail: &PassOrFail,
+) -> SourceSpan {
+    let yaml = YamlLoader::load_from_str(
+        // TODO: Should we just save the string after we read it originally?
+        yaml_text,
+    )
+    .expect("to be able to parse yaml for error reporting");
+    let YamlElt::Hash(hash) = &yaml.docs[0].yaml else {
+        unreachable!("must be a top level hashmap in the yaml")
+    };
+    let tests_hash_key = hash
+        .keys()
+        .find(|x| {
+            let YamlElt::String(str) = &x.yaml else { return false };
+            str == "tests"
+        })
+        .expect("to be able to find tests hash in yaml file");
+    let YamlElt::Hash(tests_hash) = &hash[tests_hash_key].yaml else {
+        unreachable!("there must be a tests hashmap in the yaml")
+    };
+    let pass_hash_key = tests_hash
+        .keys()
+        .find(|x| {
+            let YamlElt::String(str) = &x.yaml else { return false };
+            str == if matches!(pass_or_fail, PassOrFail::Pass) { "pass" } else { "fail" }
+        })
+        .expect("to be able to find pass hash in yaml file");
+    let YamlElt::Array(test_arr) = &tests_hash[pass_hash_key].yaml else {
+        unreachable!("there must be a pass array in the yaml")
+    };
```

Comment:
This seems like doing manual YAML deserialization and format validation. Any particular reason not to use something like serde here instead?

Reply:
We aren’t actually. We already did all that at the very beginning when we ran serde over the yaml. The reason we do this is to get the spans, which serde can’t provide, but are very helpful for anyone trying to debug which of their lint’s broke.